sklearn/linear_model/_ridge.py

```diff
@@ -228,6 +228,53 @@ def _solve_svd(X, y, alpha):
     return np.dot(Vt.T, d_UT_y).T
 
 
+def _solve_trf(X, y, alpha,
+               positive=False, max_iter=None, tol=1e-3,
+               X_offset=None, X_scale=None):
+    lsq_config = {
+        'method': 'trf',
+        'max_iter': max_iter,
+        'tol': tol,
+    }
+    n_samples, n_features = X.shape
+
+    if X_offset is None or X_scale is None:
+        X1 = sp_linalg.aslinearoperator(X)
+    else:
+        X1 = _get_rescaled_operator(X, X_offset, X_scale)
+
+    coefs = np.empty((y.shape[1], n_features), dtype=X.dtype)
+    Xa_shape = (n_samples + n_features, n_features)
+    if positive:
+        bounds = [[0] * n_features,
+                  [np.inf] * n_features]
+    else:
+        bounds = (-np.inf, np.inf)
+    sqrt_alpha = np.sqrt(alpha)
+
+    for i in range(y.shape[1]):
+        def mv(b):
+            return np.hstack([X1.matvec(b),
+                              sqrt_alpha[i] * b])
+
+        def rmv(b):
+            return X1.rmatvec(b[:n_samples]) \
+                + sqrt_alpha[i] * b[n_samples:]
+
+        Xa = sp_linalg.LinearOperator(shape=Xa_shape,
+                                      matvec=mv, rmatvec=rmv)
+        y_zeros = np.zeros(n_features, dtype=X.dtype)
+        y_column = np.hstack([y[:, i], y_zeros])
+        result = sp_optimize.lsq_linear(Xa, y_column,
+                                        bounds=bounds, **lsq_config)
+
+        if not result["success"]:
+            raise ValueError("Failed fitting using trf solver")
```

Comment:
I am not sure if we should rather not raise a `ConvergenceWarning` instead. In any case it would be great to include any additional info from `result` that could explain the cause of the convergence problem. Is conditioning a problem for this solver?

Reply:
I'm sorry. I've modified `ValueError` to `ConvergenceWarning` when success flag is `False`.
`lsq_linear` outputs `success=False` when:
- the algorithm is not able to make progress on the last iteration
- the maximum number of iterations is exceeded.